aspnetcore/mvc/models/file-uploads.md

```diff
@@ -1,195 +1,656 @@
+
+Two general approaches for uploading files are buffering and streaming.
+
+**Buffering**
+
+The entire file is read into an <xref:Microsoft.AspNetCore.Http.IFormFile>, which is a C# representation of the file used to process or save the file.
+
+> [!NOTE]
+> Any single buffered file exceeding 64 KB is moved from memory to a temp file on disk.
+
+The resources (disk, memory) used by file uploads depend on the number and size of concurrent file uploads. If an app attempts to buffer too many uploads, the site crashes when it runs out of memory or disk space. If the size or frequency of file uploads is exhausting app resources, use streaming.
+
+Buffering small files is covered in the following sections of this topic:
+
+* [Physical storage](#upload-small-files-with-buffered-model-binding-to-physical-storage)
+* [Database](#upload-small-files-with-buffered-model-binding-to-a-database)
+
+**Streaming**
+
+The file is received from a multipart request and directly processed or saved by the app. Streaming doesn't improve performance significantly&mdash;streaming reduces the demands for memory or disk space when uploading files.
+
+For more information, see the [Upload large files with streaming](#upload-large-files-with-streaming) section.
+
+### Upload small files with buffered model binding to physical storage
+
+To upload small files, use a multipart form or construct a POST request using JavaScript.
+
+The following example from the sample app demonstrates the use of a Razor Pages form to upload a single file.
+
+*Pages/BufferedSingleFileUploadPhysical.cshtml*:
+
+```cshtml
+<form enctype="multipart/form-data" method="post">
+    <dl>
+        <dt>
+            <label asp-for="FileUpload.FormFile"></label>
+        </dt>
+        <dd>
+            <input asp-for="FileUpload.FormFile" type="file">
+            <span asp-validation-for="FileUpload.FormFile"></span>
+        </dd>
+    </dl>
+    <input asp-page-handler="Upload" class="btn" type="submit" value="Upload" />
+</form>
+```
+
+The following example (*not in the sample app*) is analogous to the prior example except that it uses JavaScript to submit the form's data:
+
+```cshtml
+<form method="post" action="BufferedSingleFileUploadPhysical?handler=Upload" 
+    asp-antiforgery="true" enctype="multipart/form-data" 
+    onsubmit="AJAXSubmit(this);return false;">
+    <dl>
+        <dt>
+            <label for="FileUpload_FormFile">File</label>
+        </dt>
+        <dd>
+            <input id="FileUpload_FormFile" type="file" 
+                name="FileUpload.FormFile">
+        </dd>
+    </dl>
+
+    <input class="btn" type="submit" value="Upload">
+
+    <div style="margin-top:15px">
+        <output name="result"></output>
     </div>
 </form>
+
+@section Scripts {
+    @{await Html.RenderPartialAsync("_ValidationScriptsPartial");}
+
+    <script>
+        "use strict";
+
+        function AJAXSubmit (oFormElement) {
+          var oReq = new XMLHttpRequest();
```

Comment:
Should this use [fetch](https://caniuse.com/#search=fetch)? It's likely that if you needed to support a browser that did not support it, you're going to use a polyfill library \ jQuery.

Reply:
Yes, let's go that way, but I'm going to add this to the 3.0 updates for next week.